arbnode/resource_management.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2023, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package arbnode
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"regexp"
+	"strconv"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	flag "github.com/spf13/pflag"
+)
+
+func InitResourceManagement(conf *ResourceManagementConfig) {
+	if conf.MemoryLimitPercent > 0 {
+		node.WrapHTTPHandler = func(srv http.Handler) (http.Handler, error) {
+			return newResourceManagementHttpServer(srv, newLimitChecker(conf)), nil
+		}
+	}
+}
+
+type ResourceManagementConfig struct {
+	MemoryLimitPercent int `koanf:"mem-limit-percent" reload:"hot"`
+}
+
+var DefaultResourceManagementConfig = ResourceManagementConfig{
+	MemoryLimitPercent: 0,
+}
+
+func ResourceManagementConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Int(prefix+".mem-limit-percent", DefaultResourceManagementConfig.MemoryLimitPercent, "RPC calls are throttled if system memory utilization exceeds this percent value, zero (default) is disabled")
+}
+
+type resourceManagementHttpServer struct {
+	inner http.Handler
+	c     limitChecker
+}
+
+func newResourceManagementHttpServer(inner http.Handler, c limitChecker) *resourceManagementHttpServer {
+	return &resourceManagementHttpServer{inner: inner, c: c}
+}
+
+func (s *resourceManagementHttpServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	exceeded, err := s.c.isLimitExceeded()
+	if err != nil {
+		log.Error("Error checking memory limit", "err", err, "checker", s.c)
+	} else if exceeded {
+		http.Error(w, "Too many requests", http.StatusTooManyRequests)
+		return
+	}
+
+	log.Info("Limit not exceeded, serving request.")
+	s.inner.ServeHTTP(w, req)
+}
+
+type limitChecker interface {
+	isLimitExceeded() (bool, error)
+	String() string
+}
+
+func newLimitChecker(conf *ResourceManagementConfig) limitChecker {
```

Comment:
Return concrete types, not interfaces.

https://google.github.io/styleguide/go/decisions#interfaces: "Accept Interfaces, Return Concrete Types"

Reply:
The idea with this function is that it's a factory method that auto-discovers the type of limitChecker to create. Currently there are only 2 (cgroupsv1 and trivial). Is there a better way to structure this?

Reply:
You can do that logic (that decides the type of limitChecker) at call site itself (newLimitChecker) in InitResourceManagement. It's the only call site and this method is private so you won't have to duplicate this logic somewhere else.